### ISSUE TITLE
Fixed theta and eta

### ISSUE THREAD

**Issue by DimitrisVoul** [nlmixrdevelopment/nlmixr#294]
Hello,

Is there a way to include in nlmixr a parameter were the value of theta and eta are know and we don't want them to be estimated?

While of course estimating other covariates.

**Comment by mattfidler** [comment]
Hi @DimitrisVoul,

Yes you can fix these items. 

Here are the examples:

``` r

library(nlmixr)

## Fixing theta:

one.compartment <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- fix(3.45)    # Log V
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

fit2 <- nlmixr(one.compartment, theo_sd, est="saem", control=list(print=0))
#> Compiling RxODE equations...done.
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> Calculating residuals/tables
#> done.
print(fit2$parFixed)
#>        Parameter  Est.     SE  %RSE Back-transformed(95%CI) BSV(CV%)
#> tka       Log Ka 0.451  0.193  42.9       1.57 (1.07, 2.29)     71.1
#> tcl       Log Cl  1.02 0.0831  8.16       2.77 (2.35, 3.26)     27.0
#> tv         Log V  3.45  FIXED FIXED                    31.5     13.9
#> add.sd           0.695                                0.695         
#>        Shrink(SD)%
#> tka       -0.294%>
#> tcl         3.35%<
#> tv          10.8%<
#> add.sd


## fixing eta+theta (here is an example)
# This requires ID to be a numeric value without any gaps in ID before hand...
one.compartment <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- fix(3.45)    # Log V
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    if (id == 3){ 
      v = 4 ## fixed eta+ theta
    }
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

fit3 <- nlmixr(one.compartment, theo_sd, est="saem", control=list(print=0))
#> Compiling RxODE equations...done.
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> Needed Covariates:
#> [1] "id"
#> Calculating residuals/tables
#> done.

library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> 
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> 
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

## Here you can see ID 3 the exact value is fixed ETA+THETA
fit3 %>% filter(ID==3) %>% select(ID, v) %>% head() %>% print()
#> ── nlmixr SAEM(ODE); OBJF not calculated fit ─────────────────────────────────── 
#>  Gaussian/Laplacian Likelihoods: AIC(value) or value$objf etc. 
#>  FOCEi CWRES & Likelihoods: addCwres(value) 
#> 
#> ── Time (sec; value$time): ───────────────────────────────────────────────────── 
#>          saem    setup table covariance    other
#> elapsed 4.378 0.139274 0.008      0.014 0.285726
#> 
#> ── Population Parameters (value$parFixed or value$parFixedDf): ───────────────── 
#>        Parameter Est.     SE  %RSE Back-transformed(95%CI) BSV(CV%) Shrink(SD)%
#> tka       Log Ka   -2 0.0619  3.09    0.135 (0.119, 0.152)     16.9      14.1% 
#> tcl       Log Cl 1.25 0.0634  5.06        3.5 (3.09, 3.96)     19.2      7.36% 
#> tv         Log V 3.45  FIXED FIXED                    31.5     82.9      94.8% 
#> add.sd           1.22                                 1.22                     
#>  
#> 
#>   Covariance Type (value$covMethod): linFim
#>   No correlations in between subject variability (BSV) matrix
#>   Full BSV covariance (value$omega) 
#>     or correlation ( value $omegaR ; diagonals=SDs) 
#>   Distribution stats (mean/skewness/kurtosis/p-value) available in value$shrink 
#> 
#> ── Fit Data (object value is a modified tibble): ─────────────────────────────── 
#> # A tibble: 6 x 2
#>   ID        v
#>   <fct> <dbl>
#> 1 3         4
#> 2 3         4
#> 3 3         4
#> # … with 3 more rows

## Fixing omega variances
one.compartment <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- fix(3.45)    # Log V
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ fix(0.1)
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

fit4 <- nlmixr(one.compartment, theo_sd, est="saem", control=list(print=0))
#> Compiling RxODE equations...done.
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> Calculating residuals/tables
#> done.

print(fit4$parFixed)
#>        Parameter  Est.     SE  %RSE Back-transformed(95%CI)  BSV(CV%)
#> tka       Log Ka 0.451  0.193  42.9       1.57 (1.07, 2.29)      71.1
#> tcl       Log Cl  1.02 0.0831  8.16       2.77 (2.35, 3.26)      27.0
#> tv         Log V  3.45  FIXED FIXED                    31.5 fix(13.9)
#> add.sd           0.695                                0.695          
#>        Shrink(SD)%
#> tka       -0.294%>
#> tcl         3.35%<
#> tv          10.8%<
#> add.sd
```

<sup>Created on 2020-05-19 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
<details><summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.0 (2019-04-26)
#>  os       Gentoo/Linux                
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.utf8                  
#>  ctype    en_US.utf8                  
#>  tz       US/Central                  
#>  date     2020-05-19                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package       * version     date       lib source        
#>  assertthat      0.2.1       2019-03-21 [1] CRAN (R 3.6.0)
#>  backports       1.1.7       2020-05-13 [1] CRAN (R 3.6.0)
#>  brew            1.0-6       2011-04-13 [1] CRAN (R 3.6.0)
#>  callr           3.4.3       2020-03-28 [1] CRAN (R 3.6.0)
#>  cli             2.0.2       2020-02-28 [1] CRAN (R 3.6.0)
#>  colorspace      1.4-1       2019-03-18 [1] CRAN (R 3.6.0)
#>  crayon          1.3.4       2017-09-16 [1] CRAN (R 3.6.0)
#>  data.table      1.12.8      2019-12-09 [1] CRAN (R 3.6.0)
#>  desc            1.2.0       2018-05-01 [1] CRAN (R 3.6.0)
#>  devtools        2.3.0       2020-04-10 [1] CRAN (R 3.6.0)
#>  digest          0.6.25      2020-02-23 [1] CRAN (R 3.6.0)
#>  dparser         0.1.8       2017-11-13 [1] CRAN (R 3.6.0)
#>  dplyr         * 0.8.5       2020-03-07 [1] CRAN (R 3.6.0)
#>  ellipsis        0.3.0       2019-09-20 [1] CRAN (R 3.6.0)
#>  evaluate        0.14        2019-05-28 [1] CRAN (R 3.6.0)
#>  fansi           0.4.1       2020-01-08 [1] CRAN (R 3.6.0)
#>  fs              1.4.1       2020-04-04 [1] CRAN (R 3.6.0)
#>  generics        0.0.2       2018-11-29 [1] CRAN (R 3.6.0)
#>  ggplot2         3.3.0       2020-03-05 [1] CRAN (R 3.6.0)
#>  glue            1.4.1       2020-05-13 [1] CRAN (R 3.6.0)
#>  gtable          0.3.0       2019-03-25 [1] CRAN (R 3.6.0)
#>  highr           0.8         2019-03-20 [1] CRAN (R 3.6.0)
#>  htmltools       0.4.0       2019-10-04 [1] CRAN (R 3.6.0)
#>  huxtable        4.7.1       2020-01-08 [1] CRAN (R 3.6.0)
#>  knitr           1.28        2020-02-06 [1] CRAN (R 3.6.0)
#>  lattice         0.20-38     2018-11-04 [2] CRAN (R 3.6.0)
#>  lazyeval        0.2.2       2019-03-15 [1] CRAN (R 3.6.0)
#>  lifecycle       0.2.0       2020-03-06 [1] CRAN (R 3.6.0)
#>  lotri           0.2.2       2020-04-30 [1] local         
#>  magrittr        1.5         2014-11-22 [1] CRAN (R 3.6.0)
#>  Matrix          1.2-17      2019-03-22 [2] CRAN (R 3.6.0)
#>  memoise         1.1.0       2017-04-21 [1] CRAN (R 3.6.0)
#>  munsell         0.5.0       2018-06-12 [1] CRAN (R 3.6.0)
#>  mvnfast         0.2.5       2018-01-31 [1] CRAN (R 3.6.0)
#>  n1qn1           6.0.1-7     2020-03-19 [1] CRAN (R 3.6.0)
#>  nlme            3.1-139     2019-04-09 [2] CRAN (R 3.6.0)
#>  nlmixr        * 1.1.1-7     2020-03-18 [1] CRAN (R 3.6.0)
#>  pillar          1.4.4       2020-05-05 [1] CRAN (R 3.6.0)
#>  pkgbuild        1.0.8       2020-05-07 [1] CRAN (R 3.6.0)
#>  pkgconfig       2.0.3       2019-09-22 [1] CRAN (R 3.6.0)
#>  pkgload         1.0.2       2018-10-29 [1] CRAN (R 3.6.0)
#>  PreciseSums     0.3         2018-04-12 [1] CRAN (R 3.6.0)
#>  prettyunits     1.1.1       2020-01-24 [1] CRAN (R 3.6.0)
#>  processx        3.4.2       2020-02-09 [1] CRAN (R 3.6.0)
#>  ps              1.3.3       2020-05-08 [1] CRAN (R 3.6.0)
#>  purrr           0.3.4       2020-04-17 [1] CRAN (R 3.6.0)
#>  R6              2.4.1       2019-11-12 [1] CRAN (R 3.6.0)
#>  Rcpp            1.0.4.6     2020-04-09 [1] CRAN (R 3.6.0)
#>  RcppArmadillo   0.9.870.2.0 2020-04-27 [1] CRAN (R 3.6.0)
#>  remotes         2.1.1       2020-02-15 [1] CRAN (R 3.6.0)
#>  rex             1.2.0       2020-04-21 [1] CRAN (R 3.6.0)
#>  rlang           0.4.6       2020-05-02 [1] CRAN (R 3.6.0)
#>  rmarkdown       2.1         2020-01-20 [1] CRAN (R 3.6.0)
#>  rprojroot       1.3-2       2018-01-03 [1] CRAN (R 3.6.0)
#>  RxODE         * 0.9.2-0     2020-03-13 [1] CRAN (R 3.6.0)
#>  scales          1.1.1       2020-05-11 [1] CRAN (R 3.6.0)
#>  sessioninfo     1.1.1       2018-11-05 [1] CRAN (R 3.6.0)
#>  stringi         1.4.6       2020-02-17 [1] CRAN (R 3.6.0)
#>  stringr         1.4.0       2019-02-10 [1] CRAN (R 3.6.0)
#>  testthat        2.3.2       2020-03-02 [1] CRAN (R 3.6.0)
#>  tibble          3.0.1       2020-04-20 [1] CRAN (R 3.6.0)
#>  tidyselect      1.1.0       2020-05-11 [1] CRAN (R 3.6.0)
#>  units           0.6-6       2020-03-16 [1] CRAN (R 3.6.0)
#>  usethis         1.6.1       2020-04-29 [1] CRAN (R 3.6.0)
#>  utf8            1.1.4       2018-05-24 [1] CRAN (R 3.6.0)
#>  vctrs           0.3.0       2020-05-11 [1] CRAN (R 3.6.0)
#>  vpc             1.2.0       2020-05-07 [1] CRAN (R 3.6.0)
#>  withr           2.2.0       2020-04-20 [1] CRAN (R 3.6.0)
#>  xfun            0.13        2020-04-13 [1] CRAN (R 3.6.0)
#>  yaml            2.2.1       2020-02-01 [1] CRAN (R 3.6.0)
#> 
#> [1] /home/matt/R/x86_64-pc-linux-gnu-library/3.6
#> [2] /usr/lib64/R/library
```

</details>

Note you can also fix the values in model piping:

https://nlmixrdevelopment.github.io/nlmixr/articles/modelPiping.html#changing-and-fixing-parameter-values-in-models

**Comment by DimitrisVoul** [comment]
That's perfect thank you very much.
